packages/slate-react/src/components/string.tsx

```diff
@@ -55,15 +55,42 @@ const String = (props: {
 /**
  * Leaf strings with text in them.
  */
+type TextStringProps = { text: string; isTrailing?: boolean }
+class TextString extends React.Component<TextStringProps> {
```

Comment:
Can you keep this as a functional component and use `useEffect` and `React.memo` instead to achieve the changes?

Reply:
Not entirely sure how to get the same behavior of both (1) stopping the render, and (2) making sure React still reconciles (the `forceUpdateFlag` part), using hook mechanics.

Spent a few hours experimenting, and was able to get (1) with a React.memo wrap on a subcomponent (could probably be lifted up to `<Leaf />`) and not updating render value if `textContent` is the same as `text` prop, but haven't been able to solve (2).

(2) Manifests itself as seeing ghost characters on delete (e.g. typing `hi|`, delete, expect to see `h|`, but see `h|i`) or selection moving around (`hello |`, delete, `|hello`.

I'm a little stumped, but also haven't quite grokked functional components enough to think of a clever way of doing it. Happy to give it another go if anyone's got any ideas.

Reply:
@luddep @ianstormtaylor The problem is that you want to prevent a normal render unless you want to reconcile the textContent with the native element, so it is not so simple to do it with hooks. I think you will need to have two hook based components to do this, something like the following (sorry if I mangled the TS a little bit)?

```
const TextString = (props: { text: string; isTrailing?: boolean }) => {
  const { text, isTrailing = false } = props

  const ref = useRef(null);
  const forceUpdateFlag = useRef(false);

  if (ref.current && ref.current.textContent !== text) {
    forceUpdateFlag.current = !forceUpdateFlag.current;
  }

  return <TextStringMemo ref={ref} text={text} isTrailing={isTrailing} forceUpdate={forceUpdateFlag.current} />
}

const TextStringMemo = memo(forwardRef((props: { text: string; isTrailing: boolean, forceUpdate: boolean }, ref) => {
  return (
    <span
      data-slate-string
      ref={ref}
      key={forceUpdate ? 'A' : 'B'}
    >
      {text}	
      {isTrailing ? '\n' : null}
    </span>
  )
}));
```

Reply:
Actually, if I understand what you are trying to do is just make sure that the render is forced on text change it should be possible with a single component after you make it `memo`:

```
const TextString = React.memo((props: { text: string; isTrailing?: boolean }) => {
  const { text, isTrailing = false } = props

  const ref = useRef(null);
  const forceUpdateFlag = useRef(false);

  // Always change the key when a change to the text field, which will force react to render
  forceUpdateFlag.current = !forceUpdateFlag.current;

  return (
    <span
      data-slate-string
      ref={ref}
      key={forceUpdateFlag.current ? 'A' : 'B'}
    >
      {text}	
      {isTrailing ? '\n' : null}
    </span>
  );
});
```